wordpress-mobile/WordPress-iOS
# Stats: Improve scrolling performance

Fixes #22721

Majority of Stats Traffic / Stats Period scrolling hitches happen due to inefficient `TopTotalsCell` load, mainly calls to `addRows` which creates and adds multiple layers of `UIStackViews` every time the cell is configured.

Investigation:  pc8eDl-1pe-p2

## Solution

- Create and add `UIStackView`  only once when the table view loads. When other cells reuse the same views, only configure or hide existing views.
- Use `setNeedsLayout` instead of `layoutIfNeeded`

This is only the first iteration of possible improvements. I didn't achieve a full smooth experience and animation hitches still happen after the first load. However, the improvements are measurable and visible, especially when testing on a debugger-connected iPhone.

## Performance Comparisons

I decided to compare the performance of cell configuration methods by measuring how much time it takes to fully configure the cell. When testing, I use a high-traffic site and scroll up and down three times.

### TopTotalsCell

#### iPhone 14 Pro

| Condition | Performance Average (s) | Improvement Factor |
|-----------|-------------------------|--------------------|
| Before    | 0.023552464503867953    | -                  |
| After     | 0.005567020840115017    | 4.2x faster        |

### CountriesCell

#### iPhone 14 Pro

| Condition | Performance Average (s) | Improvement Factor |
|-----------|-------------------------|--------------------|
| Before    | 0.02454930543899536     | -                  |
| After     | 0.004590545381818499   | 5.3x faster       |

### All Stats Cells Averages

The improvement of the configuration of all the Stats cell averages falls somewhere in between of two numbers.

#### iPhone 14 Pro

| Condition | Performance Average (s) | Improvement Factor |
|-----------|-------------------------|--------------------|
| Before    | 0.011049999985643613    | -                  |
| After     | 0.0023589220599851747   | 4.68x faster       |


## Hitches Comparisons

### Before

<img width="2137" alt="Hitches Before - No Debugger" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/d5015996-63ba-4359-b863-840e94a79176">

### After

<img width="1885" alt="Hitches After - With Subtitle Visibility" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/eb0eb271-a4dd-4a2d-be81-963c4191017e">

## Remaining Hitches

The majority of those remaining hitches are caused by issues that couldn't be identified or sometimes identified simply as `main` using Time Profiler:

<img width="580" alt="image" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/8b32bc38-07eb-4987-be18-97a9b2e17425">

## To test:

1. Build Jetpack app on `trunk` with a debugger connected
2. Select higher-traffic site
3. Open Stats, select Traffic tab
4. Scroll up and down a couple of times, and observe the performance
5. Build Jetpack app on `task/22721-stats-traffic-improve-scrolling-performance` with a debugger connected
6. Open Stats, select Traffic tab
7. Scroll up and down a couple of times, and observe the performance
8. Confirm that cell appearance is not broken
9. Disconnect the debugger
10. Open Stats, select Traffic tab
11. Scroll up and down a couple of times, and observe the performance
12. Confirm that cell appearance is not broken

## Regression Notes
1. Potential unintended areas of impact

TopTotalsCell is only used within Stats, doesn't affect any other areas

2. What I did to test those areas of impact (or what existing automated tests I relied on)

Manual testing, and performance measurement

3. What automated tests I added (or what prevented me from doing so)

None

PR submission checklist:


Testing checklist:
